File: tsp/solvers/branch_and_bound.py
"""
Branch and Bound Solver

Note:
  This file includes a python bnb solver, but also adds an interface to a cpp
  implementation. The default is the cpp version.
"""

from collections import defaultdict
import copy
import heapq
import math
import numpy as np
from numba.experimental import jitclass
from numba import uint32, jit, njit
from numba.np.extensions import cross2d
import time
from heapq import heappush, heappop

from _solvers import branch_and_bound as bnb_cpp

@njit
def prim_find_mst(disjoint_items, dist_mat):
    """
    Calculates minimum spanning tree weight using Prim's algorithm
    """
    N = len(disjoint_items)
    cost = 0.0
    tree = Bitset(N)
    start = 0
    queue = [(np.uint32(0), start)]

    while queue:
        weight, vtx = heappop(queue)
        if not tree.contains(vtx):
            tree.add(vtx)
            cost += weight
            for j in range(N):
                neigh = disjoint_items[j]
                if not tree.contains(j):
                    heappush(queue, (dist_mat[disjoint_items[vtx], neigh], j))
    return cost


# For Bitset class below
spec = [
    ('_N', uint32),
    ('_bitsize', uint32),
    ('_bitset', uint32[:]),
    ('_len', uint32)
    ]

@jitclass(spec)
class Bitset:
    """
    Set datastructure with a small memory footprint to hold nonnegative integers 0...N-1
    
    Used for determining remaining nodes in the tour
    """
    def __init__(self, N: int):

        self._N = N
        self._bitsize = 32
        array_len = math.ceil(self._N / self._bitsize)
        self._bitset = np.zeros(array_len, dtype=np.dtype('uint32'))
        self._len = 0

    def len(self):
        return self._len
        
    def contains(self, item):
        if item < self._N:
            block_idx, elem_idx = divmod(item, self._bitsize)
            return (self._bitset[block_idx] & (1 << elem_idx)) > 0
        else:
            return False

    def add(self, item):
        """Add item to set"""
        if item < self._N:
            block_idx, elem_idx = divmod(item, self._bitsize)
            self._bitset[block_idx] |= (1 << elem_idx)
            self._len += 1

    def remove(self, item):
        """Remove item from set"""
        if item < self._N:
            block_idx, elem_idx = divmod(item, self._bitsize)
            self._bitset[block_idx] &= ~(1 << elem_idx)
            self._len -= 1

    def empty(self):
        for i in range(len(self._bitset)):
            if self._bitset[i] != 0:
                return False
        return True

    def npitems(self):
        arr = np.zeros(self._len, dtype=np.dtype('uint32'))
        j = 0
        for i in range(self._N):
            if self.contains(i):
                arr[j] = i
                j = j + 1
        return arr

    def items(self):
        return [i for i in range(self._N) if self.contains(i)]

    def copy(self):
        b = Bitset(self._N)
        b._bitset = self._bitset.copy()
        b._len = self._len
        return b
    

class Node:
    """
    Branch and bound node
    """
    
    def __init__(self, item, subproblem: Bitset, level=0, parent=None, distance=0):
        self.item = item
        self.subproblem = subproblem
        self.level = level
        self.parent = parent
        self.distance = distance
        self.lower_est = 0

        self.cost = float('inf')

    def __lt__(self, other):
        return self.priority < other.priority

    @property
    def priority(self):
        """Prioritize leaf nodes with a smaller lower bound"""
        return self.lowerbound / self.level
    
    @property
    def path(self):
        """
        Returns the path of the tour up until this node
        """
        if not hasattr(self, '_path'):
            path = []
            node = self
        
            while node is not None:
                path.append(node.item)
                node = node.parent
            
            self._path = list(reversed(path))
        return self._path

    def expand(self, adj_mat):
        """
        Iterator over each node in the current subproblem
        """
        level = self.level + 1
        for item in self.subproblem.items():
            new_subproblem = self.subproblem.copy()
            new_subproblem.remove(item)
            distance = self.distance + adj_mat[self.item, item]
            yield Node(
                item, new_subproblem, level=level, distance=distance, parent=self
            )

    @property
    def lowerbound(self):
        """
        Lower bound is current distance plus lower estimation on remaining nodes
        """
        return self.distance + self.lower_est

    def set_cost(self, cost):
        """
        Sets the total cost (should be done once a tour is complete).
        Cost is equal to the distance for a full tour, which includes traveling back to start.
        """
        self.cost = cost

    def set_lower_est(self, lower_est):
        """
        Set lower estimation on remaining nodes
        """
        self.lower_est = lower_est

@njit
def direction(a, b, c):
    """
    Find direction by computing cross product.
    Used for intersection code.
    """
    return cross2d(c-a, b-a)

@njit
def intersect(a, b, c, d):
    """
    Determine if line segments (a,b) and (c,d) intersect.
    Allow segments that share endpoints.
    """
    d1 = direction(c,d,a)
    d2 = direction(c,d,b)
    d3 = direction(a,b,c)
    d4 = direction(a,b,d)
    if (((d1 > 0 and d2 < 0) or (d1 < 0 and d2 > 0))
        and ((d3 > 0 and d4 < 0) or (d3 < 0 and d4 > 0))):
        return True
    return False
        
def branch_and_bound_py(tsp_data, max_time, debug=False, **kwargs):
    """
    Banch and bound solver
    """

    start_time = time.time()
    adj_mat = tsp_data.to_adjacency_mat()

    # Initialize unvisited vertices
    all_candidates = Bitset(len(adj_mat))
    for i in range(1, len(adj_mat)):
        all_candidates.add(i)

    # Level of terminal leaf nodes
    max_level = len(adj_mat) - 1

    # Priority queue holding most promising nodes
    F = [Node(0, all_candidates)]
    best_solution = F[0]

    idx = 0
    trace = []

    def no_intersections(node):
        """
        Helper code for determining if adding this node 
        results in an intersection among edges in tour
        """

        # If only 1 or 2 nodes in tour, then no intersection possible
        if node.parent is None or node.parent.parent is None:
            return True
        
        cur_edge_b = tsp_data.coords[node.item]
        cur_edge_a = tsp_data.coords[node.parent.item]

        # Loop through each previous edge in tour
        node = node.parent
        b = tsp_data.coords[node.item]
        while node.parent is not None:
            a = tsp_data.coords[node.parent.item]
            if intersect(a, b, cur_edge_a, cur_edge_b):
                return False

            node = node.parent
            b = a

        return True

    # Main branch-and-bound loop
    while F and (time.time()-start_time) < max_time:
        
        node = heapq.heappop(F)

        # Expand subproblem
        for subnode in node.expand(adj_mat):
            # Solution found
            if subnode.level == max_level:                
                subnode.set_cost(subnode.distance + adj_mat[subnode.item,0])
                # Best solution?
                if subnode.cost < best_solution.cost:
                    best_solution = subnode
                    trace.append([best_solution.cost, time.time()-start_time])
            else:
                # Optimal solution doesn't have intersecting edges
                if not no_intersections(subnode):
                    continue

                # Calculate lower estimation on remaining nodes
                lower_est = 0
                if not subnode.subproblem.empty():

                    disjoint_items = subnode.subproblem.npitems()
                    
                    # 1. Shortest edges from start and current node
                    from_start = np.min(adj_mat[0,disjoint_items])
                    from_curr = np.min(adj_mat[subnode.item, disjoint_items])

                    lower_est += from_start + from_curr

                    # 2. Minimum spanning tree of remaining elements
                    cost = prim_find_mst(disjoint_items, adj_mat)
                    lower_est += int(cost)
                    
                subnode.set_lower_est(lower_est)

                # Not a dead end, add node to queue
                if subnode.lowerbound < best_solution.cost:
                    if debug:
                        print(subnode.level, subnode.lowerbound, best_solution.cost, idx, int(time.time()-start_time))
                    heapq.heappush(F, subnode)
                    idx += 1

    return best_solution.cost, best_solution.path, trace


def branch_and_bound_cpp(tsp_data, max_time, depth_first=False, debug=False):
    """
    Interface to cpp branch-and-bound implementation
    """
    solution = bnb_cpp(tsp_data.coords, max_time, depth_first, debug)
    best_tour = solution.tour
    best_dist = solution.distance

    trace = []
    for trace_item in solution.trace:
        trace.append([trace_item.distance, trace_item.time])

    return best_dist, best_tour, trace


def branch_and_bound(tsp_data, max_time, debug=False, lang='cpp'):
    """
    Forwards args to selected language's implementation
    """
    if lang == 'cpp':
        return branch_and_bound_cpp(tsp_data, max_time, debug=debug)
    elif lang == 'py':
        return branch_and_bound_py(tsp_data, max_time, debug=debug)

    
class BranchAndBound:
    """
    Solver interface
    """
    def __init__(self, tsp_data, max_time, debug=False, lang='cpp'):
        self.tsp_data = tsp_data
        self.max_time = max_time
        self.debug = debug
        self.lang = lang

    def solve(self):
        return branch_and_bound(self.tsp_data, self.max_time, self.debug, self.lang)
    

if __name__ == '__main__':

    import sys
    import time
    
    from tsp.parse import parse

    if (len(sys.argv) < 2):
        print("Usage: branch_and_bound.py <city>")
        exit(0)
    
    filename = f"../DATA/{sys.argv[1]}.tsp"
    d = parse(filename)
    
    sol = branch_and_bound(d, 600)
    print(sol)
    
    
    

File: tsp/solvers/src/branch_and_bound.cpp
/**
 * Branch and bound solver
 */
#include <algorithm>
#include <chrono>
#include <iostream>
#include <queue>
#include <set>
#include <vector>
#include "branch_and_bound.h"

#include <pybind11/pybind11.h>

namespace BnB {
    
DistanceMatrix::DistanceMatrix(const std::vector<Coord>& coords) :
    m_size(coords.size())
{
    m_dist_mat.reserve(m_size);
    
    for (size_t i = 0; i < coords.size(); ++i) {
	for (size_t j = 0; j < coords.size(); ++j) {
	    uint32_t dist = euc_2d(coords[i], coords[j]);
	    m_dist_mat.push_back(dist);
	}
    }
}

TspNode::TspNode(Index index, std::vector<bool>&& subproblem, Index level, std::shared_ptr<TspNode> parent, uint32_t distance) :
	index(index), level(level), subproblem(std::move(subproblem)), parent(parent), distance(distance), lower_est(0), cost(UINT_MAX)
{
}

std::vector<std::shared_ptr<TspNode>>
TspNode::expand(const DistanceMatrix& dist_mat)
{
    std::vector<std::shared_ptr<TspNode>> subnodes;
    uint32_t new_level = level + 1;
    
    for (uint32_t i = 0; i < subproblem.size(); ++i) {
	if (subproblem[i]) {
	    std::vector<bool> new_subproblem = subproblem;
	    new_subproblem[i] = false;
	    uint32_t new_distance = distance + dist_mat.at(index, i);
	    	    
	    std::shared_ptr<TspNode> new_node(new TspNode{i, std::move(new_subproblem), new_level, shared_from_this(), new_distance});
	    	    
	    subnodes.push_back(new_node);
	}
    }
    return subnodes;
}
    
void
TspNode::update_lower_est(const DistanceMatrix& dist_mat)
{
    lower_est = 0;
    
    // Create vector of remaining nodes
    std::vector<Index> remaining_nodes;
    
    for (uint32_t i = 0; i < subproblem.size(); ++i) {
	if (subproblem[i]) {
	    remaining_nodes.push_back(i);

	}
    }

    // Shortest edges from start and current node
    uint32_t from_start = std::numeric_limits<uint32_t>::max();
    uint32_t from_curr = std::numeric_limits<uint32_t>::max();
    for (const auto& i : remaining_nodes) {
	from_start = std::min(from_start, dist_mat.at(0, i));
	from_curr = std::min(from_curr, dist_mat.at(index, i));
    }
    if (!remaining_nodes.empty())
	lower_est += from_start + from_curr;

    // Sum of minimum spanning tree of remaining nodes
    std::set<uint32_t> tree;  // mst
    
    // Use a priority queue
    using qnode = std::pair<uint32_t, uint32_t>;
    class qnode_cmp
    {
    public:
	bool operator() (const qnode& lhs, const qnode&rhs) const
	    {
		return lhs.first > rhs.first;
	    }
    };
    typedef std::priority_queue<qnode, std::vector<qnode>, qnode_cmp> tree_queue_type;
    
    tree_queue_type tree_queue;
    tree_queue.push({0, 0});

    while (!tree_queue.empty()) {
	auto [weight, idx] = tree_queue.top();
	tree_queue.pop();

	if (tree.count(idx) == 0) {
	    tree.insert(idx);
	    lower_est += weight;  // update mst weight
	    for (size_t i = 0; i < remaining_nodes.size(); ++i) {
		if (tree.count(i) == 0) {
		    auto neighbor = remaining_nodes[i];
		    tree_queue.push({dist_mat.at(remaining_nodes[idx], neighbor), i});
		}
	    }
	}
    }
}

std::vector<Index>
TspNode::path() const
{
    std::vector<Index> _path;
    _path.push_back(index);
    
    auto node_ptr = parent;
    while (node_ptr != nullptr) {
	_path.push_back(node_ptr->index);
	node_ptr = node_ptr->parent;
    }
    std::reverse(_path.begin(), _path.end());
    return _path;
}
        
Solution solve(const std::vector<Coord>& coords, float max_time, bool depth_first, bool debug)
{
    auto start_time = std::chrono::high_resolution_clock::now();

    // Helper function for determining if adding node causes intersections with
    // any edge in path
    auto no_intersections = [&](const std::shared_ptr<TspNode>& node) {

	if (node->parent == nullptr or node->parent->parent == nullptr) {
	    return true;
	}

	Coord cur_b = coords[node->index];
	Coord cur_a = coords[node->parent->index];

	auto node_ptr = node->parent;
	Coord b = coords[node_ptr->index];
	while (node_ptr->parent != nullptr) {
	    Coord a = coords[node_ptr->parent->index];
	    if (intersect(a, b, cur_a, cur_b)) {
		return false;
	    }
	    
	    node_ptr = node_ptr->parent;
	    b = a;
	}

	return true;
    };

    // Helper function for getting current time since start
    auto get_time = [=]() -> double {
        using namespace std::chrono;	

	auto time_now = high_resolution_clock::now();
	duration<double> ts = time_now - start_time;
	return ts.count();
    };

    // Adjacency matrix
    DistanceMatrix dist_mat(coords);

    // Initialize unvistited nodes
    std::vector<bool> all_candidates(coords.size());
    all_candidates.flip();
    all_candidates[0] = 0; // Except first node

    // Level of terminal leaf nodes
    uint32_t max_level = coords.size() - 1;

    // Initialize priority queue holding nodes in branch and bound solver
    std::vector<std::shared_ptr<TspNode>> node_queue;
    node_queue.push_back(std::shared_ptr<TspNode>(new TspNode(0, std::move(all_candidates))));

    // Initialize solution with closest neighbor
    auto node_ptr = node_queue.front();
    for (size_t l = 0; l < max_level; ++l) {
	auto best_subnode = node_ptr;
	uint32_t best_dist = UINT_MAX;
	for (const auto& subnode : node_ptr->expand(dist_mat)) {
	    if (subnode->distance < best_dist) {
		best_subnode = subnode;
		best_dist = subnode->distance;
	    }
	}
	node_ptr = best_subnode;
    }
    std::shared_ptr<TspNode> best_solution = node_ptr;
    best_solution->cost = best_solution->distance + dist_mat.at(best_solution->index, 0);
    
    std::vector<TraceItem> trace;
    trace.push_back({best_solution->cost, get_time()});

    // Main solver loop
    while (!node_queue.empty() && (get_time() < max_time)) {

	// Check for ctrl-C on python side
	if (PyErr_CheckSignals() != 0) {
	    if (debug) {
		std::cout << "-- Caught keyboard interrupt --\n";
	    }
	    break;
	}
	
	// Pop off top node
	auto node = node_queue.back();
	if (!depth_first) {
	    node = node_queue.front();
	    std::pop_heap(node_queue.begin(), node_queue.end(), NodeCmp());
	}
	node_queue.pop_back();

	// Expand subproblem
	for (auto& subnode : node->expand(dist_mat)) {

	    // Found a solution (at terminal leaf node)
	    if (subnode->level == max_level) {

		// Total distance includes traveling back to beginning
		subnode->cost = subnode->distance + dist_mat.at(subnode->index, 0);

		// Best solution found?
		if (subnode->cost < best_solution->cost) {
		    best_solution = subnode;
		    trace.push_back({best_solution->cost, get_time()});
		}
	    }
	    else {
		// Optimal solutions contain no intersections
		if (!no_intersections(subnode)) {
		    continue;
		}

		// Update lower estimation of remaining nodes (MST lower bound
		// functions)
		subnode->update_lower_est(dist_mat);

		// Not a dead end?
		if (subnode->lowerbound() < best_solution->cost) {
		    node_queue.push_back(subnode);
		    if (!depth_first)
			std::push_heap(node_queue.begin(), node_queue.end(), NodeCmp());

		    if (debug) {
			std::cout << subnode->level << ", "
				  << subnode->lowerbound() << ", "
				  << best_solution->cost << ", "
				  << get_time() << ", "
				  << node_queue.size() << '\n';
		    }
		}
	    }
	}
    }

    if (debug) {
	std::cout << "Finished\n";
	std::cout << "  " << best_solution->cost << '\n';
    }
    
    return {best_solution->path(), best_solution->cost, trace};
}

} // namespace
